### ISSUE TITLE
How to lock down the RethinkDB "admin" account?

### ISSUE THREAD

**Issue by ttmc** [bigchaindb/bigchaindb#240]
@killerstorm pointed out that anyone with admin access to RethinkDB could just drop the database. What can be done to prevent that and similar things (e.g. dropping tables)?

Each node owner could get a RethinkDB user account with the minimum necessary permissions. That should suffice for day-to-day operations (e.g. reading from and writing to certain tables).

There's still an "admin" account. (See [the RethinkDB docs on permissions and user accounts](https://rethinkdb.com/docs/permissions-and-accounts/).) It can't be deleted. It doesn't have a password by default, but it can be given one.

One can change the admin password by updating the `users` system table. Who has permission to do that? Anyone with `update` permissions on the `users` table. Maybe we could make it so only that admin user has that permission?

Maybe we could throw away the password for the admin user, so nobody can do admin-level things. Maybe not: what if we really need to do admin-level things? What?

The RethinkDB web administration UI assumes that whoever is using it is admin, and it never asks for a password, so maybe it should be set so _nobody_ can access to the web UI, not even the admin of a RethinkDB server. (Otherwise whoever has access can just change the admin password by updating the `users` table.) Is that even possible? How?

Maybe we can fork RethinkDB and remove the web UI. But what's to prevent a rogue node from running standard RethinkDB (with web UI)?

Can we put a changefeed on the `users` table, so that all nodes will be notified if it changes? Even if other nodes find out, is there anything they can do about it? Maybe they have to revert to a recent backup?

Are there other ways to change the admin password or otherwise get admin-level access?
### Why this Matters

We _want_ to build a decentralized system, where all admin-type proposals must be decided upon (and enacted) by voting. How do we get that using RethinkDB (or any similar database system), with its admin user?

If we can lock down the admin user, its password could be encrypted by multiple passwords, and it could be decrypted as-needed, to do specific already-voted-upon tasks under supervision.

Maybe it's possible to do away with the admin user completely, by forgetting its password and preventing it from getting a new one. Nodes can monitor the database settings to detect any un-approved changes.

Maybe the admin user can be allowed to exist as usual, but every node must check _all_ RethinkDB wire protocol traffic, only allowing a short whitelist of allowed operations (and raising alarms if it sees requests for disallowed operations). The whitelist can be created by voting.

**Comment by ttmc** [comment]
From the RethinkDB docs:

"Note: As of version 2.3, only the `admin` user can access system tables. Read [Permissions and user accounts](https://www.rethinkdb.com/docs/permissions-and-accounts/) for more details on user accounts and permissions."

Source:  https://www.rethinkdb.com/docs/system-tables/

**Comment by ttmc** [comment]
A copy of what was written for #598, which duplicated this issue:

Transcript of the sticky note:

DB admin worked out
- initialization
- live
Every action becomes a "request for action"

(I think what this means is that if anyone wants to make a database-admin-level change to the database, they must first do a request to the whole federation and then all the other nodes must vote or whatever.)

This is closely related to cluster configuration management (issue #436) and the wire protocol firewall (issue #594).

**Comment by ttmc** [comment]
Our current thinking on this is that during initial database setup, the admin user will be used to set up the database tables, secondary indexes, sharding, replication, etc. Then the wire protocol firewall (issue #594) will be turned on and after that it will be up to the wire protocol firewall to accept/deny all admin-type actions. The federation governance may decide to allow some limited set of admin actions for a limited amount of time, as needed; and this would be enforced by the wire protocol firewall.

In other words, this issue is mostly subsumed by the wire protocol firewall issue, so I'm closing it.

**Comment by ttmc** [comment]
Our thinking is that the wire protocol firewall will be used to handle this. (Those will probably be turned off during database setup, but once they are turned on [one per node], it will take a federation-level decision to change the firewall rules to allow certain admin actions to work.)

In other words, this issue is mostly subsumed by the wire protocol firewall issue (#594), so I'm closing it.
